Add `screenshot/{2,3}` function for taking a screenshot manually, and `screenshot: true` config for auto-capture

s3cur3 opened the pull request:
Resolves #8

s3cur3 commented on lib/phoenix_test/playwright/frame.ex:
```diff
@@ -24,6 +24,30 @@ defmodule PhoenixTest.Playwright.Frame do
+  def screenshot(frame_id, path, opts \\ []) do
+    params =
+      opts
+      |> Keyword.validate!(
+        # Playwright options: https://playwright.dev/docs/api/class-page#page-screenshot
+        full_page: true,
+        omit_background: false
```
I tried adding support for other keys (like `:quality` on JPGs), but Playwright seemed to either ignore them or just fail to produce a screenshot entirely. I figure we can revisit in the future to add support for more of Playwright's options.

s3cur3 commented on lib/phoenix_test/playwright/serialization.ex:
```diff
@@ -63,4 +63,27 @@ defmodule PhoenixTest.Playwright.Serialization do
   defp to_atom(s) when is_binary(s), do: String.to_atom(s)
   defp to_atom(a) when is_atom(a), do: a
+
+  def camel_case_keys(enum) when is_map(enum) or is_list(enum) do
```
I'm not sure this is entirely in the spirit of the library, but I find it more natural to write `|> screenshot("foo.png", full_page: true)` rather than `fullPage: true`, even though Playwright documents the option in the camelCase version.

ftes replied:
I agree with you. But I would rather do this globally for all messages we send to playwright to keep things consistent.

So remove from this PR and move e.g. to `Port`.
Similar to this: https://github.com/mechanical-orchard/playwright-elixir/blob/4d0f740e972066310fa1ca133e3ae6ffab06b1ef/lib/playwright/sdk/channel/message.ex#L36

ftes replied:
#11 should do that if you would care to have a look.

ftes replied:
Done

ftes commented on test/phoenix_test/playwright_test.exs:
```diff
@@ -12,6 +12,31 @@ defmodule PhoenixTest.PlaywrightTest do
+      :ok
+    end
+
+    test "takes a screenshot of the current page as a PNG", %{conn: conn} do
```
Probably coincidental: Other tests are now failing even in retry run with `--max-cases 1`

ftes reviewed (commented):
I'd prefer to switch to auto-camel-casing globally instead of just for screenshots.

s3cur3 commented:
Okay, I rebased onto `main` now that #11 is in, and added a `:screenshot` config option to automatically take a screenshot in an `on_exit` handler (a88eb6b). I think it should be good to go!

ftes commented:
Sorry for sitting on this for a while.
I was thinking config handling in this library might be improved (not just this PR but in general).
But that doesn't have to happen right now.

Thanks!

ftes reviewed (approved):
Sorry for sitting on this for a while.
I was thinking config handling in this library might be improved (not just this PR but in general).
But that doesn't have to happen right now.

Thanks!

ftes commented on .github/workflows/elixir.yml:
```diff
@@ -62,10 +62,23 @@ jobs:
         run: "mix test --warnings-as-errors --max-cases 1 || if [[ $? = 2 ]]; then PLAYWRIGHT_TRACE=true mix test --max-cases 1 --failed; else false; fi"
+      - name: Fail if screenshot on exit failed
+        run: |
+          if ! ls screenshots_for_verification/test_screenshot_3_saves_screenshot_on_test_exit* >/dev/null 2>&1; then
```
Neat test!

ftes commented:
FYI I moved `Frame.screenshot` to `Page.screenshot` and hoisted the file handling into `Case`.
The idea is to keep the `Frame/Page` modules as a slim wrapper/documentation of the Playwright API.
